pkg/remote/remote.go

```diff
@@ -147,45 +152,54 @@ func authOptions(ctx context.Context, ref name.Reference, option types.RemoteOpt
 	}
 }
 
-func parsePlatform(ref name.Reference, p string, options []remote.Option) (*v1.Platform, error) {
+// resolvePlatform resolves the OS platform for a given image reference.
+// If the platform has an empty OS, the function will attempt to find the first OS
+// in the image's manifest list and return the platform with the detected OS.
+// It ignores the specified platform if the image is not multi-arch.
+func resolvePlatform(ref name.Reference, p *types.Platform, options []remote.Option) (*types.Platform, error) {
+	if p.OS != "" {
+		return p, nil
+	}
+
 	// OS wildcard, implicitly pick up the first os found in the image list.
 	// e.g. */amd64
-	if strings.HasPrefix(p, "*/") {
-		d, err := remote.Get(ref, options...)
-		if err != nil {
-			return nil, xerrors.Errorf("image get error: %w", err)
-		}
-		switch d.MediaType {
-		case v1types.OCIManifestSchema1, v1types.DockerManifestSchema2:
-			// We want an index but the registry has an image, not multi-arch. We just ignore "--platform".
-			log.Logger.Debug("Ignore --platform as the image is not multi-arch")
-			return nil, nil
-		case v1types.OCIImageIndex, v1types.DockerManifestList:
-			// These are expected.
-		}
-
-		index, err := d.ImageIndex()
-		if err != nil {
-			return nil, xerrors.Errorf("image index error: %w", err)
-		}
+	d, err := remote.Get(ref, options...)
+	if err != nil {
+		return nil, xerrors.Errorf("image get error: %w", err)
+	}
+	switch d.MediaType {
+	case v1types.OCIManifestSchema1, v1types.DockerManifestSchema2:
+		// We want an index but the registry has an image, not multi-arch. We just ignore "--platform".
+		log.Logger.Debug("Ignore --platform as the image is not multi-arch")
```

Comment:
This code only considers multi-arch images..What about "force" platform check in single arch images?
e.g: arm64v8/alpine:3.9.3
Below are the scenarios that we tested using trivy OSS
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>
<head>
	
	
	<title></title>
	<meta name="generator" content="LibreOffice 6.4.7.2 (Linux)"/>
	<style type="text/css">
		body,div,table,thead,tbody,tfoot,tr,th,td,p { font-family:"Liberation Sans"; font-size:x-small }
		a.comment-indicator:hover + comment { background:#ffd; position:absolute; display:block; border:1px solid black; padding:0.5em;  } 
		a.comment-indicator { background:red; display:inline-block; border:1px solid black; width:0.5em; height:0.5em;  } 
		comment { display:none;  } 
	</style>
	
</head>

<body>


Scenario | Actual | Expected
-- | -- | --
  |   |  
-d image --list-all-pkgs --server http://localhost:8085 --platform "windows/arm64" alpine:3.9.3 (multi-arch image) | Scan is passing even though the image’s manifest  doesn’t have windows OS | Scan should fail saying OS is not available or platform didn’t match
  |   |  
-d image --list-all-pkgs --server http://localhost:8085 --platform "*/amd64" arm64v8/alpine:3.9.3 (single arch image) | Scan is passing even though the image’s  config (not image manifest) is not list type and has only one arch which is arm64.. doesn’t have amd64 arch | Scan should fail saying architecture didn’t match
  |   |  
-d image --list-all-pkgs --server http://localhost:8085 --platform "*/arm64" arm64v8/alpine:3.9.3 (single arch image) | Scan is passing | Scan should pass because the provided arch and imge’s arch matches and “*” should be replaced by “linux” or resp OS
  |   |  
-d image --list-all-pkgs --server http://localhost:8085 --platform "windows/arm64" arm64v8/alpine:3.9.3 (single arch image) | Scan is passing even though the image’s config (not image manifest)  doesn’t have windows OS | Scan should fail saying OS is not available..image’s config is linux
  |   |  
-d image --list-all-pkgs --server http://localhost:8085 --platform "windows/amd64" alpine:3.9.3 (multi-arch image) | Scan is passing even though the image’s manifest  doesn’t have windows OS | Scan should fail saying OS didn’t match..image’s OS is linux.none of the manifests has this OS/arch combo
  |   |  
-d image --list-all-pkgs --server http://localhost:8085 --platform "linux/amd64" arm64v8/alpine:3.9.3 (single arch image) | Scan is passing even though the image’s config (not image manifest) is not list type and has only one arch which is arm64.. doesn’t have amd64 arch | Scan should fail saying architecture didn’t match


</body>

</html>

Reply:
Why do we want to care about single-arch images? If I remember correctly, we will give a list of platforms, and users will choose it.

Reply:
I confirmed the above cases worked properly.